






                                                                        Brady Gehrman
                                                                      CSE 625 Project 1
                                                                  Date Submitted: 6/7/2023












Problem 1: In this problem, you will deal with the issues related to adding up a large collect of
 small numbers

In [3]:
eps32 = eps(0f0)

1.0f-45

The value return when running eps(0f0) was 1.0f-45. The eps function returns the numerical gap (epsilon value) between the value of the input parameter and the next larger value that can be represented within Julia's floating-point system. In this case the input parameter to the eps function is 0f0, a 32-bit floating point number in Julia, and the function returned 1.0f-45, which is the epsilon value for 0f0, meaning that 0f0 + 1.0f-45 would yeild you the next floating point number after 0f0.

In [9]:
eps64 = eps(0.0)


5.0e-324

The value return when running eps(0.0) was 5.0e-324. The eps function returns the numerical gap (epsilon value) between the value of the input parameter and the next value that can be represented within Julia's floating-point system. In this case the input parameter to the eps function is 0.0, a 64-bit floating point number in Julia, and the function returned 5.0e-324, which is the epsilon value for 0.0, meaning that 0.0 + 5.0e-324 would yeild you the next floating point number after 0.0.

In [4]:
v = fill(eps32, 2_000_000_000) # this requires 8G bytes of memory space)

2000000000-element Vector{Float32}:
 1.0f-45
 1.0f-45
 1.0f-45
 1.0f-45
 1.0f-45
 1.0f-45
 1.0f-45
 1.0f-45
 1.0f-45
 1.0f-45
 ⋮
 1.0f-45
 1.0f-45
 1.0f-45
 1.0f-45
 1.0f-45
 1.0f-45
 1.0f-45
 1.0f-45
 1.0f-45

In [15]:
typeof(v)

Vector{Float32} (alias for Array{Float32, 1})

In [19]:
"""
using for loop to find the sum of v
"""
sumForLoop = 0f0
@time for value in v
    sumForLoop += value
end
sumForLoop

217.921912 seconds (8.00 G allocations: 149.012 GiB, 4.34% gc time)


2.3509887f-38

In [8]:
"""
using foldr function to find the sum of v
"""

@time foldr(+, v)

  3.164249 seconds (4 allocations: 64 bytes)


2.3509887f-38

In [24]:
"""
using reduce function to find the sum of v
"""

@time reduce(+, v)

  0.224165 seconds (1 allocation: 16 bytes)


2.802597f-36

In [25]:
"""
using sum function to find the sum of v
"""

@time sum(v)


  0.216971 seconds (1 allocation: 16 bytes)


2.802597f-36

In [9]:
v_float64 = Float64.(v)

2000000000-element Vector{Float64}:
 1.401298464324817e-45
 1.401298464324817e-45
 1.401298464324817e-45
 1.401298464324817e-45
 1.401298464324817e-45
 1.401298464324817e-45
 1.401298464324817e-45
 1.401298464324817e-45
 1.401298464324817e-45
 1.401298464324817e-45
 ⋮
 1.401298464324817e-45
 1.401298464324817e-45
 1.401298464324817e-45
 1.401298464324817e-45
 1.401298464324817e-45
 1.401298464324817e-45
 1.401298464324817e-45
 1.401298464324817e-45
 1.401298464324817e-45

In [12]:
sum_float64 = sum(v_float64)

2.802596928649634e-36

In [11]:
foldr(+, v_float64)

2.802596928649634e-36

Compare the results and run-times of computing sum of v using these four methods:

1 for loop:
run-time: 217.921912 seconds,
result: 2.3509887f-38


2 the foldr function:
run-time: 3.164249 seconds,
result: 2.3509887f-38


3 the reduce function:
run-time: 0.224165 seconds,
result: 2.802597f-36


4 the sum function:
run-time: 0.216971 seconds,
result: 2.802597f-36

Fastest-to-Slowest: sum function, reduce function, foldr function, then the for loop immplementation

Explaination on why some of the results are accurate and some are not: 

The reason some of the results are accurate and some are not is because of floating-point precision and potential rounding errors. Vector v holds elements of type Float32, as shown above in my notebook. The for loop and the foldr function seem to produce the same result (2.3509887f-38), while the reduce and sum functions both produce a different result (2.802597f-36). As we all know Float32 addition is not as accurate or precise as Float64 addition when it comes to very small numbers such as the ones being added in this scenerio; therefore, I then created a new vector representation of vector v, but each of the its elements were converted to Float64 elements. I then found the sum of this new array to be 2.802596928649634e-36 via the sum function and the foldr function, leading me to beleive that the sum and reduce immplementations produced more accurate results the first time around.




Problem 2: In this problem, you will compare different implementations of Fibonacci numbers.

In [13]:
fib_1(n) = first([1 1;1 0]^(n-1)*[1,0])
fib_2(n) = last( [1 1;1 0]^(n-1)*[1,1])

fib_2 (generic function with 1 method)

In [15]:
@time fib_1(40)

  0.000018 seconds (11 allocations: 1024 bytes)


102334155

In [16]:
@time fib_2(40)

  0.000007 seconds (11 allocations: 1024 bytes)


102334155

Justify why the two Julia functions (fib_1 and fib_2) work:

The function fib_1 and fib_2 calculate the fibonacci number using matrix exponentitation. The nth fibonacci number can be calculated by raising the matrix [1 1; 0 0] to the power of (n-1), and then extracting the first or last element of the matrix result. This process works because each previous fibonacci sequence is stored inside of the resultant matrix each time it is multiplied.

In [17]:
"""
Recursive Julia function for Fibonacci
"""
function fib_recursive(n)
    if n <= 0
        println("input parameter must be positive")
    elseif n == 1 || n == 2
        return 1
    else
        return fib_recursive(n-1) + fib_recursive(n-2)
    end
end

@time fib_recursive(40)

  1.577256 seconds


102334155

The Time Performance comparision of the 3 Fibonacci immplementations (Average execution times based on multiple tests)

Trial 1: n = 10
fib_1: 0.000023 seconds,
fib_2: 0.000005 seconds,
fib_re: 0.000008 seconds

Trial 2: n = 20
fib_1: 0.000043 seconds,
fib_2: 0.000008 seconds,
fib_r: 0.000124 seconds

Trial 3: n = 30
fib_1: 0.000012 seconds,
fib_2: 0.000009 seconds,
fib_r: 0.015140 seconds

Trial 4: n = 40
fib_1: 0.000019 seconds,
fib_2: 0.000013 seconds,
fib_r: 1.575460 seconds

Trial 5: n = 50
fib_1: 0.000007 seconds,
fib_2: 0.000010 seconds,
fib_r: 200.933958 seconds

Time Comparison:

Fastest Function: fib_2 (O(log N))

2nd Fastest Function: fib_1 (O(log N))

Slowest: fib_recursive (O(2^n))

The matrix immplementations of the fibonacci are very similar in execution time, but the fib_2 function was faster for each trial. The recursive immplementation was comparable to the matrix immplementations when n was relativley small, but as n is increased, the execution time of the recursive function increased very quickly compared to matrix immplemenations.


Problem 3: 

Consider the C function of implementing sin function with the Taylor series:
 void sinx(int N, int terms, float* x, float* result)
 {
    for (int i = 0; i < N; i++)
    {
        float value = x[i];
        float numer = x[i] * x[i] * x[i];
        float denom = 6; // 3!
        float sign = -1;
        for (int j = 1; j <= terms; j++)
        {
            value += sign * numer / denom;
            numer *= x[i] * x[i];
            denom *= (float)(2 * j + 2) * (2 * j + 3);
            sign *= -1;
        }
        result[i] = value;
    }
 }

In [1]:
"""
Implement the C function in a Julia function with the following specification:

sinx implements sin function with the Taylor series.
Arguments:
x angle
n number of terms used
degree x in degree if degree is true else in radians
"""

function sinx_forLoop(x, n = 5; degree = true)
    if degree
        x = deg2rad(x)
    end
    result = x
    sign = 1

    for i in 1:n-1
        sign *= -1
        numerator = sign * x^(2i+1)
        denominator =  factorial(2*i + 1)
        result += numerator / denominator
    end
    return result
end

function sinx(x, n = 5; degree = true)
    if degree
        x = deg2rad(x)
    end

    terms = map(i -> ((-1)^(i-1) * x^(2i-1)) / factorial(2i-1), 1:n)
    result = reduce(+, terms)

    return result
end

println(sinx_forLoop(90))
println(sinx(90))

1.0000035425842861


1.0000035425842861


In [27]:
"""
Try to find a small n for sinx so that the absolute differences of their values for all degrees in the range 0:1:90 are less than 0.1e-7
"""
function findBestNValue()
    for n in 1:15
        valid = true
        for i in 0:1:90
            if(abs(sind(i) - sinx(i, n)) > 0.00000001)
                valid = false
            end
        end
        if valid
            return n
        end
    end
end

findBestNValue()

7

The smallest n for my sinx function so that the absolute difference between its values and Julia's bulit-in sind function's values is less than 0.1e-7 is 7 based on the output of the function that I created above! 

In [5]:
"""
Print sinx function in a pretty table for x in the degree range 0:10:90
"""
function printSineTable()
    print("│ ")
    printstyled("Degree(s)", color=:cyan)
    print(" │ ")
    printstyled(rpad(lpad("Sinx", 8), 11), color=:cyan)
    print(" │\n")
    for i in 0:10:90
        sineValue = sinx(i)
        print("│ ")
        print(lpad(i, 5), "     │",
        rpad(lpad(round(sineValue, digits=10), 2), 12), " │\n")
    end
end

printSineTable()

│ Degree(s) │     Sinx    │
│     0     │0.0          │
│    10     │0.1736481777 │
│    20     │0.3420201433 │
│    30     │0.5          │
│    40     │0.6427876102 │
│    50     │0.7660444487 │
│    60     │0.8660254451 │
│    70     │0.9396928454 │
│    80     │0.9848087259 │
│    90     │1.0000035426 │


In [29]:
"""
Use the macro @time to measure the run-time of your sinx (n = 10) and the built-in sind functions to compute sin(x) for angle x in the degree range of
0:inc:90, for inc = 0.1, 0.01, 0.001, 0.000_1, 0.000_01, 0.000_001, and 0.000_000_1, respectively. Assemble your timing results in a table. Discuss your timing results
"""

using Printf

function findTimeForSinx(increment)
    [sinx(i, 10) for i in 0:increment:90]
end

function findTimeForSind(increment)
    [sind(i) for i in 0:increment:90]
end

function createTimingTable()
    increments = [0.1, 0.01, 0.001, 0.000_1, 0.000_01, 0.000_001, 0.000_000_1]
    @printf("| %s | %s | %s |\n", "Increment", "Sinx (s)", "Sind (s)")
    for inc in increments
        timeForSinx = @elapsed findTimeForSinx(inc)
        timeForSind = @elapsed findTimeForSind(inc)
        timeForSinx = round(timeForSinx, digits=3)
        timeForSind = round(timeForSind, digits=3)
        @printf("| %f  | %f | %f |\n", inc, timeForSinx, timeForSind)
    end
end

createTimingTable()

| Increment | Sinx (s) | Sind (s) |


| 0.100000  | 0.098000 | 0.000000 |
| 0.010000  | 0.052000 | 0.000000 |


| 0.001000  | 0.180000 | 0.001000 |


| 0.000100  | 1.648000 | 0.019000 |


| 0.000010  | 16.561000 | 0.188000 |


| 0.000001  | 176.117000 | 2.578000 |


| 0.000000  | 2005.395000 | 54.753000 |


Discussion of Timing Results:

As can be seen my outputted table above, my sinx function is much slower than Julia's built-in sind function. This can especially be seen when the increment size decreases to very small amounts, meaning we needed to run the sinx and sind function more times. Julia's sind function has certainly been further optimized than my sinx function. 
